book/src/rfcs/foreign_field_add.md

```diff
@@ -201,40 +201,48 @@ with $r_{top} = r_2$ and $c = c_1$.
-|             |                                  |         |
-| 9n+8        | `ForeignFieldAdd`                |         |
-| 9n+9        | `Zero`                           |         |
+The full foreign field addition/subtraction gadget will be composed by: 
+- $1$ public input row containing the value $1$;
+- $n$ rows with `ForeignFieldAdd` gate type:
+  - for the actual $n$ chained additions or subtractions;  
+- $1$ `ForeignFieldAdd` row for the final bound check;
+- $1$ `Zero` row for the final upper bound check.
+- $1$ `RangeCheck` gadget for the first left input of the chain $a_1 := a$;
+- Then, $n$ of the following set of rows:
+  - $1$ `RangeCheck` gadget for the $i$-th right input $b_i$;
+  - $1$ `RangeCheck` gadget for the $i$-th result which will correspond to the $(i+1)$-th left input of the chain $r_i = a_{i+1}$.
+- $1$ final `RangeCheck` gadget for the upper bound term $u$.
+
+A total of 20 rows with 15 columns in Kimchi for 1 addition. All ranges below are inclusive.
+
+| Row(s)           | Gate type(s)        | Witness |
+| ---------------- | ------------------- | ------- |
+| 0                | `PublicInput`       | $1$     |
+| 1..n             | `ForeignFieldAdd`   | +/-     |
+| n+1              | `ForeignFieldAdd`   | bound   |
+| n+2              | `Zero`              | bound   |
+| n+3..n+6         | `multi-range-check` | left    |
+| n+7+8i..n+10+8i  | `multi-range-check` | right   |
+| n+11+8i..n+14+8i | `multi-range-check` | $r$     |
+| 9n+7..9n+10      | `multi-range-check` | $u$     |
+
+
+This mechanism can chain foreign field additions together. Initially, there are $n$ foreign field addition gates, followed by a foreign field addition gate for the upper bound (whose current row corresponds to the next row of the last foreign field addition gate), and an auxiliary `Zero` row that holds the upper bound. At the end, an initial left input range check is performed, which is followed by a $n$ pairs of range check gates for the right input and intermediate result (which become the left input for the next iteration). After the chained inputs checks, a final range check on the upper bound takes place. 
+
+Nonetheless, such an exhaustive set of checks are not necessary for completeness nor soundness. In particular, only the very final range check for the bound is required. Thus, a shorter gadget that is equally valid and takes $(8*n+4)$ less rows is possible, and it follows the next layout (with inclusive ranges):
+
+| Row(s)   | Gate type(s)                                          | Witness |
+| -------- | ----------------------------------------------------- | ------- |
+| 0        | public input row for soundness of bound overflow flag | $1$     |
+| 1..n     | `ForeignFieldAdd`                                     |         |
+| n+1      | `ForeignFieldAdd`                                     |         |
+| n+2      | `Zero`                                                |         |
```

Comment:
But okay, I guess if the very initial inputs were not real limbs, the field overflow could happen regardless.

Reply:
My understanding was that the argument for allowing $\ge f$ inputs is restricted to the case when the input is the result of a previous addition.  That is, the inputs to addition are either the result of a previous addition or else something smaller than $f$.  Otherwise, without a range check on the actual values summed, I could start with an input equal to the max native modulus (i.e. $n - 1$) and trigger an overflow without hitting the max circuit size.  I think we should err on the side of caution and check with @mrmr1993 .

Reply:
Well, this is kind of collateral to FFAdd, seems like we will have to address an efficient way when we look at the real use case indeed

Reply:
Ah, so you think it's orthogonal?  Yeah, I guess we will build a custom layout with the flexible interface anyway and then the gadget creator can always constrain those if needed.

Still someone could use this gadget an end up with something possibly under-constrained, no?

Reply:
I mean, right now the builder allows you to create a chain of additions, but any other check you may want to add needs to be included in a larger gadget. FFAdd specs warn about the needed checks you may want, so when we write a larger gadget we will be able to have plenty of flexibility on where to put each gate. If it is too monolithic, then it may not be useful for ecdsa as we may want to put things in between to chain with multiplications in the middle, etc.